v5/core/parameterized_url.go

```diff
@@ -0,0 +1,83 @@
+package core
+
+// (C) Copyright IBM Corp. 2021.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+//
+// constructServiceURL returns a service URL that is constructed by formatting a parameterized URL.
+//
+// Parameters:
+//
+// parameterizedUrl: URL that contains variable placeholders, e.g. "{scheme}://ibm.com".
+//
+// defaultUrlVariables: map from variable names to default values.
+//		Each variable in the parameterized URL must have a default value specified in this map.
+//
+// providedUrlVariables: map from variable names to desired values.
+// 		If a variable is not provided in this map,
+// 		the default variable value will be used instead.
+//
+func constructServiceURL(
+	parameterizedUrl string,
+	defaultUrlVariables map[string]string,
+	providedUrlVariables map[string]string,
+) (string, error) {
+
+	// If null was passed, we set the variables to an empty map.
+	// This results in all default variable values being used.
+	if providedUrlVariables == nil {
+		providedUrlVariables = map[string]string{}
+	}
+
+	// Verify the provided variable names.
+	for providedName := range providedUrlVariables {
+		if _, ok := defaultUrlVariables[providedName]; !ok {
+			// Get all variable names (the keys of the default variables map).
+			acceptedNames := make([]string, len(defaultUrlVariables))
+
+			i := 0
+			for name := range defaultUrlVariables {
```

Comment:
I think it would be cleaner to move the creation of the "acceptedNames" slice before this for loop on line 50, then combine that for loop together with the second one on line 73.
IOW, compute the list of valid variable names ahead of time, then enter a loop where you are formatting the URL AND checking for invalid variable names in one step.